File: src/gobby/code_index/indexer.py
"""Code indexer orchestrator.

Coordinates: parse -> store -> embed -> graph
for individual files and full directories.
"""

from __future__ import annotations

import logging
import subprocess
import time
from collections.abc import Callable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from gobby.code_index.chunker import chunk_file_content
from gobby.code_index.graph import CodeGraph
from gobby.code_index.hasher import file_content_hash
from gobby.code_index.languages import detect_language, get_extensions_map
from gobby.code_index.models import (
    IndexedFile,
    IndexedProject,
    IndexResult,
    Symbol,
)
from gobby.code_index.parser import CodeParser
from gobby.code_index.security import should_exclude
from gobby.code_index.storage import CodeIndexStorage
from gobby.config.code_index import CodeIndexConfig

logger = logging.getLogger(__name__)


def _get_git_files(root: Path) -> set[str] | None:
    """Get git-visible files (tracked + untracked-but-not-ignored).

    Returns relative paths, or None if not a git repo / git unavailable.
    """
    try:
        # Tracked files
        result = subprocess.run(
            ["git", "ls-files"],
            cwd=str(root),
            capture_output=True,
            text=True,
            timeout=30,
        )
        if result.returncode != 0:
            return None

        files: set[str] = set()
        for line in result.stdout.strip().splitlines():
            if line:
                files.add(line)

        # Untracked but not ignored
        result2 = subprocess.run(
            ["git", "ls-files", "--others", "--exclude-standard"],
            cwd=str(root),
            capture_output=True,
            text=True,
            timeout=30,
        )
        if result2.returncode == 0:
            for line in result2.stdout.strip().splitlines():
                if line:
                    files.add(line)

        return files
    except (OSError, subprocess.TimeoutExpired):
        return None


class CodeIndexer:
    """Orchestrates code indexing: parse -> store -> embed -> graph."""

    def __init__(
        self,
        storage: CodeIndexStorage,
        parser: CodeParser,
        vector_store: Any | None = None,
        embed_fn: Callable[..., Any] | None = None,
        graph: CodeGraph | None = None,
        config: CodeIndexConfig | None = None,
    ) -> None:
        self._storage = storage
        self._parser = parser
        self._vector_store = vector_store
        self._embed_fn = embed_fn
        self._graph = graph
        self._config = config or CodeIndexConfig()
        # Cache collections known to be missing to avoid repeated failed HTTP roundtrips
        self._missing_collections: set[str] = set()

    @property
    def storage(self) -> CodeIndexStorage:
        return self._storage

    @property
    def graph(self) -> CodeGraph | None:
        return self._graph

    @property
    def config(self) -> CodeIndexConfig:
        return self._config

    async def index_directory(
        self,
        root_path: str,
        project_id: str,
        incremental: bool = True,
    ) -> IndexResult:
        """Index all supported files in a directory.

        Args:
            root_path: Absolute path to directory root.
            project_id: Project identifier.
            incremental: If True, skip unchanged files.

        Returns:
            IndexResult with counts and timing.
        """
        start = time.monotonic()
        result = IndexResult(project_id=project_id)
        root = Path(root_path).resolve()

        if not root.is_dir():
            result.errors.append(f"Not a directory: {root_path}")
            return result

        ext_map = get_extensions_map()
        supported_extensions = set(ext_map.keys())
        content_extensions = set(self._config.content_extensions)
        all_extensions = supported_extensions | content_extensions

        # Collect candidate files using git-aware discovery
        git_files = _get_git_files(root)
        candidates: list[Path] = []
        content_only_candidates: list[Path] = []

        if git_files is not None:
            # Git-aware: only index files git knows about
            for rel_path in git_files:
                path = root / rel_path
                if not path.is_file():
                    continue
                suffix = path.suffix.lower()
                if suffix not in all_extensions:
                    continue
                if should_exclude(path, self._config.exclude_patterns):
                    continue
                if suffix in supported_extensions:
                    candidates.append(path)
                else:
                    content_only_candidates.append(path)
        else:
            # Fallback: rglob for non-git repos
            for path in root.rglob("*"):
                if not path.is_file():
                    continue
                suffix = path.suffix.lower()
                if suffix not in all_extensions:
                    continue
                if should_exclude(path, self._config.exclude_patterns):
                    continue
                if suffix in supported_extensions:
                    candidates.append(path)
                else:
                    content_only_candidates.append(path)

        # Build hash map for incremental check
        current_hashes: dict[str, str] = {}
        if incremental:
            for path in candidates:
                try:
                    rel = str(path.resolve().relative_to(root))
                    current_hashes[rel] = file_content_hash(path)
                except (OSError, ValueError):
                    pass

            stale = set(self._storage.get_stale_files(project_id, current_hashes))
        else:
            stale = None  # Index everything

        # Clean up orphan files no longer in candidates (e.g., newly excluded dirs)
        if incremental and current_hashes:
            try:
                orphans = self._storage.get_orphan_files(project_id, set(current_hashes.keys()))
                for orphan_path in orphans:
                    await self._delete_file_data(project_id, orphan_path)
                if orphans:
                    logger.info(f"Cleaned up {len(orphans)} orphan indexed files")
            except Exception as e:
                logger.debug(f"Orphan cleanup failed: {e}")

        # Index each file (symbols + content chunks)
        for path in candidates:
            try:
                rel = str(path.resolve().relative_to(root))
            except ValueError:
                continue

            if incremental and stale is not None and rel not in stale:
                result.files_skipped += 1
                continue

            symbols = await self.index_file(str(path), project_id, root_path)
            if symbols is not None:
                result.files_indexed += 1
                result.symbols_found += len(symbols)
            else:
                result.files_skipped += 1

        # Index content-only files (no AST parsing, just content chunks)
        for path in content_only_candidates:
            try:
                rel = str(path.resolve().relative_to(root))
            except ValueError:
                continue
            await self._index_content_only(path, project_id, root_path)

        # Update project stats
        elapsed_ms = int((time.monotonic() - start) * 1000)
        result.duration_ms = elapsed_ms

        total_files = self._storage.count_files(project_id)
        total_symbols = self._storage.count_symbols(project_id)

        self._storage.upsert_project_stats(
            IndexedProject(
                id=project_id,
                root_path=root_path,
                total_files=total_files,
                total_symbols=total_symbols,
                last_indexed_at=datetime.now(UTC).isoformat(),
                index_duration_ms=elapsed_ms,
            )
        )

        logger.info(
            f"Indexed {result.files_indexed} files ({result.files_skipped} skipped), "
            f"{result.symbols_found} symbols in {elapsed_ms}ms"
        )
        return result

    async def _delete_file_data(self, project_id: str, file_path: str) -> None:
        """Remove all data for a file from SQLite, Qdrant, and Neo4j."""
        self._storage.delete_symbols_for_file(project_id, file_path)
        self._storage.delete_file(project_id, file_path)
        self._storage.delete_content_chunks_for_file(project_id, file_path)
        self._storage.delete_imports_for_file(project_id, file_path)
        self._storage.delete_calls_for_file(project_id, file_path)

        if self._graph is not None and self._graph.available:
            await self._graph.delete_file(file_path=file_path, project_id=project_id)

        if self._vector_store is not None:
            collection = f"{self._config.qdrant_collection_prefix}{project_id}"
            if collection not in self._missing_collections:
                try:
                    await self._vector_store.delete(
                        filters={"file_path": file_path, "project_id": project_id},
                        collection_name=collection,
                    )
                except Exception as e:
                    err_str = str(e).lower()
                    if "not found" in err_str or "doesn't exist" in err_str:
                        self._missing_collections.add(collection)
                        logger.debug(
                            f"Qdrant collection {collection} not found, "
                            f"skipping future deletes this run"
                        )
                    else:
                        logger.debug(f"Vector delete failed for {file_path}: {e}")

    async def index_file(
        self,
        file_path: str,
        project_id: str,
        root_path: str,
    ) -> list[Symbol] | None:
        """Index a single file. Returns symbols or None if skipped."""
        parse_result = self._parser.parse_file(file_path, project_id, root_path)

        root = Path(root_path).resolve()
        path = Path(file_path).resolve()
        try:
            rel_path = str(path.relative_to(root))
        except ValueError:
            rel_path = str(path)

        # Always clear old data first before re-indexing to prevent ghosts
        await self._delete_file_data(project_id, rel_path)

        if parse_result is None:
            return None

        symbols = parse_result.symbols
        if not symbols:
            return []

        # Store symbols
        self._storage.upsert_symbols(symbols)

        language = detect_language(file_path) or "unknown"
        try:
            h = file_content_hash(file_path)
            size = path.stat().st_size
        except OSError:
            h = ""
            size = 0

        self._storage.upsert_file(
            IndexedFile(
                id=IndexedFile.make_id(project_id, rel_path),
                project_id=project_id,
                file_path=rel_path,
                language=language,
                content_hash=h,
                symbol_count=len(symbols),
                byte_size=size,
            )
        )

        # Store import/call relations in SQLite (for daemon sync worker)
        self._storage.upsert_imports(project_id, rel_path, parse_result.imports)
        self._storage.upsert_calls(project_id, rel_path, parse_result.calls)

        # Index content chunks for full-text content search
        try:
            source = path.read_bytes()
            chunks = chunk_file_content(
                source=source,
                rel_path=rel_path,
                project_id=project_id,
                language=language,
            )
            if chunks:
                self._storage.upsert_content_chunks(chunks)
        except Exception as e:
            logger.debug(f"Content chunk indexing failed for {file_path}: {e}")

        file_id = IndexedFile.make_id(project_id, rel_path)

        # Embed symbols (async, non-blocking on failure)
        if self._vector_store is not None and self._embed_fn is not None:
            try:
                count = await self._embed_symbols(symbols, project_id)
                if count > 0:
                    self._storage.mark_vectors_synced(file_id)
            except Exception as e:
                logger.debug(f"Embedding failed for {file_path}: {e}")

        # Add graph relationships (async, non-blocking on failure)
        if self._graph is not None and self._graph.available:
            try:
                count = await self._add_graph_data(project_id, rel_path, parse_result, symbols)
                if count > 0:
                    self._storage.mark_graph_synced(file_id)
            except Exception as e:
                logger.debug(f"Graph update failed for {file_path}: {e}")

        return symbols

    async def index_changed_files(
        self,
        project_id: str,
        root_path: str,
        file_paths: list[str],
    ) -> IndexResult:
        """Index specific changed files (e.g., from git post-commit)."""
        start = time.monotonic()
        result = IndexResult(project_id=project_id)

        for fp in file_paths:
            # Resolve relative to root
            abs_path = Path(root_path) / fp if not Path(fp).is_absolute() else Path(fp)
            if not abs_path.exists():
                # File was deleted — clean up everywhere
                await self._delete_file_data(project_id, fp)
                continue

            symbols = await self.index_file(str(abs_path), project_id, root_path)
            if symbols is not None:
                result.files_indexed += 1
                result.symbols_found += len(symbols)

        result.duration_ms = int((time.monotonic() - start) * 1000)
        return result

    async def _index_content_only(self, path: Path, project_id: str, root_path: str) -> None:
        """Index a non-tree-sitter file for content search only."""
        root = Path(root_path).resolve()
        resolved = path.resolve()
        try:
            rel_path = str(resolved.relative_to(root))
        except ValueError:
            return

        try:
            size = path.stat().st_size
            if size == 0 or size > self._config.max_file_size_bytes:
                return
        except OSError:
            return

        try:
            source = path.read_bytes()
            # Skip binary files
            if b"\x00" in source[:8192]:
                return
        except OSError:
            return

        # Clear old chunks
        self._storage.delete_content_chunks_for_file(project_id, rel_path)

        chunks = chunk_file_content(
            source=source,
            rel_path=rel_path,
            project_id=project_id,
            language=path.suffix.lstrip(".") or None,
        )
        if chunks:
            self._storage.upsert_content_chunks(chunks)

    async def invalidate(self, project_id: str) -> None:
        """Clear all index data for a project."""
        self._storage.delete_symbols_for_project(project_id)
        self._storage.delete_files_for_project(project_id)
        self._storage.delete_content_chunks_for_project(project_id)

        # Clear graph
        if self._graph is not None:
            await self._graph.clear_project(project_id)

        # Clear vectors
        if self._vector_store is not None:
            collection = f"{self._config.qdrant_collection_prefix}{project_id}"
            try:
                await self._vector_store.delete_collection(collection)
            except Exception as e:
                logger.debug(f"Vector collection delete failed: {e}")

        logger.info(f"Invalidated code index for project {project_id}")

    async def _embed_symbols(self, symbols: list[Symbol], project_id: str) -> int:
        """Embed symbol text into Qdrant. Returns count embedded."""
        if not symbols or self._embed_fn is None or self._vector_store is None:
            return 0

        # Build text for embedding: name + signature + docstring
        texts = []
        ids = []
        for sym in symbols:
            parts = [sym.qualified_name]
            if sym.signature:
                parts.append(sym.signature)
            if sym.docstring:
                parts.append(sym.docstring[:200])
            texts.append(" ".join(parts))
            ids.append(sym.id)

        # Generate embeddings
        try:
            embeddings = []
            for text in texts:
                emb = await self._embed_fn(text)
                if emb is not None:
                    embeddings.append(emb)
                else:
                    embeddings.append(None)
        except Exception as e:
            logger.debug(f"Embedding generation failed: {e}")
            return 0

        # Upsert to Qdrant
        collection = f"{self._config.qdrant_collection_prefix}{project_id}"
        count = 0
        try:
            items = []
            for i, emb in enumerate(embeddings):
                if emb is not None:
                    items.append(
                        (
                            ids[i],
                            emb,
                            {
                                "name": symbols[i].name,
                                "kind": symbols[i].kind,
                                "file_path": symbols[i].file_path,
                                "project_id": project_id,
                            },
                        )
                    )

            if items:
                await self._vector_store.batch_upsert(
                    items=items,
                    collection_name=collection,
                )
                count = len(items)
                # Collection exists now — clear from missing cache
                self._missing_collections.discard(collection)
        except Exception as e:
            logger.debug(f"Vector upsert failed: {e}")

        return count

    async def _add_graph_data(
        self,
        project_id: str,
        rel_path: str,
        parse_result: Any,
        symbols: list[Symbol],
    ) -> int:
        """Add parsed relationships to Neo4j graph."""
        if self._graph is None:
            return 0

        imports = [
            {
                "source_file": imp.source_file,
                "target_module": imp.target_module,
            }
            for imp in parse_result.imports
        ]

        calls = [
            {
                "caller_symbol_id": call.caller_symbol_id,
                "callee_name": call.callee_name,
                "file_path": call.file_path,
                "line": call.line,
            }
            for call in parse_result.calls
        ]

        contains = [{"id": sym.id, "name": sym.name, "kind": sym.kind} for sym in symbols]

        return await self._graph.add_relationships(
            project_id=project_id,
            file_path=rel_path,
            imports=imports,
            calls=calls,
            contains=contains,
        )

File: rust/gcode/src/index/indexer.rs
//! Full and incremental indexing orchestrator.
//! Ports logic from src/gobby/code_index/indexer.py.

use std::collections::HashMap;
use std::path::Path;
use std::time::Instant;

use rusqlite::Connection;

use crate::index::chunker;
use crate::index::hasher;
use crate::index::languages;
use crate::index::parser;
use crate::index::walker;
use crate::models::{IndexResult, IndexedFile, IndexedProject};
use crate::neo4j::Neo4jClient;

/// Default exclude patterns (matching Python CodeIndexConfig defaults).
const DEFAULT_EXCLUDES: &[&str] = &[
    "node_modules", "__pycache__", ".git", ".venv", "venv", "dist", "build",
    ".tox", ".mypy_cache", ".pytest_cache", ".ruff_cache", "target",
    ".next", ".nuxt", "coverage", ".cache",
];

/// Index a directory (full or incremental).
pub fn index_directory(
    conn: &Connection,
    root_path: &Path,
    project_id: &str,
    incremental: bool,
    neo4j: Option<&Neo4jClient>,
) -> anyhow::Result<IndexResult> {
    let start = Instant::now();
    let mut result = IndexResult {
        project_id: project_id.to_string(),
        files_indexed: 0,
        files_skipped: 0,
        symbols_found: 0,
        errors: Vec::new(),
        duration_ms: 0,
    };

    let excludes: Vec<String> = DEFAULT_EXCLUDES.iter().map(|s| s.to_string()).collect();
    let (candidates, content_only) = walker::discover_files(root_path, &excludes);

    // Build current hash map for incremental detection
    let mut current_hashes: HashMap<String, String> = HashMap::new();
    let stale: Option<std::collections::HashSet<String>> = if incremental {
        for path in &candidates {
            if let Ok(rel) = relative_path(path, root_path) {
                if let Ok(h) = hasher::file_content_hash(path) {
                    current_hashes.insert(rel, h);
                }
            }
        }
        Some(get_stale_files(conn, project_id, &current_hashes))
    } else {
        None
    };

    // Clean orphans
    if incremental && !current_hashes.is_empty() {
        let orphans = get_orphan_files(conn, project_id, &current_hashes);
        for orphan in &orphans {
            delete_file_data(conn, project_id, orphan, neo4j);
        }
    }

    // Index each candidate file
    for path in &candidates {
        let rel = match relative_path(path, root_path) {
            Ok(r) => r,
            Err(_) => continue,
        };

        if let Some(ref stale_set) = stale {
            if !stale_set.contains(&rel) {
                result.files_skipped += 1;
                continue;
            }
        }

        match index_file(conn, path, project_id, root_path, &excludes, neo4j) {
            Some(count) => {
                result.files_indexed += 1;
                result.symbols_found += count;
            }
            None => {
                result.files_skipped += 1;
            }
        }
    }

    // Index content-only files
    for path in &content_only {
        index_content_only(conn, path, project_id, root_path);
    }

    let elapsed_ms = start.elapsed().as_millis() as u64;
    result.duration_ms = elapsed_ms;

    // Update project stats
    let total_files = count_rows(conn, "code_indexed_files", project_id);
    let total_symbols = count_rows(conn, "code_symbols", project_id);

    upsert_project_stats(
        conn,
        &IndexedProject {
            id: project_id.to_string(),
            root_path: root_path.to_string_lossy().to_string(),
            total_files,
            total_symbols,
            last_indexed_at: iso_now(),
            index_duration_ms: elapsed_ms,
        },
    );

    eprintln!(
        "Indexed {} files ({} skipped), {} symbols in {}ms",
        result.files_indexed, result.files_skipped, result.symbols_found, elapsed_ms
    );

    Ok(result)
}

/// Index specific changed files.
pub fn index_files(
    conn: &Connection,
    root_path: &Path,
    project_id: &str,
    file_paths: &[String],
    neo4j: Option<&Neo4jClient>,
) -> anyhow::Result<IndexResult> {
    let start = Instant::now();
    let mut result = IndexResult {
        project_id: project_id.to_string(),
        files_indexed: 0,
        files_skipped: 0,
        symbols_found: 0,
        errors: Vec::new(),
        duration_ms: 0,
    };

    let excludes: Vec<String> = DEFAULT_EXCLUDES.iter().map(|s| s.to_string()).collect();

    for fp in file_paths {
        let abs = if Path::new(fp).is_absolute() {
            std::path::PathBuf::from(fp)
        } else {
            root_path.join(fp)
        };

        if !abs.exists() {
            // File deleted — clean up
            delete_file_data(conn, project_id, fp, neo4j);
            continue;
        }

        match index_file(conn, &abs, project_id, root_path, &excludes, neo4j) {
            Some(count) => {
                result.files_indexed += 1;
                result.symbols_found += count;
            }
            None => {}
        }
    }

    result.duration_ms = start.elapsed().as_millis() as u64;
    Ok(result)
}

/// Index a single file. Returns symbol count or None if skipped.
fn index_file(
    conn: &Connection,
    file_path: &Path,
    project_id: &str,
    root_path: &Path,
    exclude_patterns: &[String],
    neo4j: Option<&Neo4jClient>,
) -> Option<usize> {
    let rel = relative_path(file_path, root_path).ok()?;

    // Clear old data first
    delete_file_data(conn, project_id, &rel, neo4j);

    let parse_result = parser::parse_file(file_path, project_id, root_path, exclude_patterns)?;

    if parse_result.symbols.is_empty() {
        return Some(0);
    }

    let count = parse_result.symbols.len();

    // Upsert symbols to SQLite
    upsert_symbols(conn, &parse_result.symbols);

    // Write graph edges to Neo4j
    if let Some(client) = neo4j {
        crate::neo4j::write_defines(client, project_id, &rel, &parse_result.symbols);
        crate::neo4j::write_calls(client, project_id, &parse_result.calls);
        crate::neo4j::write_imports(client, project_id, &parse_result.imports);
    }

    // Upsert file record
    let language = languages::detect_language(&file_path.to_string_lossy()).unwrap_or("unknown");
    let h = hasher::file_content_hash(file_path).unwrap_or_default();
    let size = file_path.metadata().map(|m| m.len()).unwrap_or(0);

    upsert_file(
        conn,
        &IndexedFile {
            id: IndexedFile::make_id(project_id, &rel),
            project_id: project_id.to_string(),
            file_path: rel.clone(),
            language: language.to_string(),
            content_hash: h,
            symbol_count: count,
            byte_size: size as usize,
            indexed_at: iso_now(),
        },
    );

    // Content chunks
    if let Ok(source) = std::fs::read(file_path) {
        let chunks = chunker::chunk_file_content(&source, &rel, project_id, Some(language));
        if !chunks.is_empty() {
            upsert_content_chunks(conn, &chunks);
        }
    }

    Some(count)
}

/// Index content-only file (no AST, just chunks).
fn index_content_only(conn: &Connection, path: &Path, project_id: &str, root_path: &Path) {
    let rel = match relative_path(path, root_path) {
        Ok(r) => r,
        Err(_) => return,
    };

    let meta = match path.metadata() {
        Ok(m) if m.len() > 0 && m.len() <= 10 * 1024 * 1024 => m,
        _ => return,
    };

    let source = match std::fs::read(path) {
        Ok(s) => s,
        Err(_) => return,
    };

    // Skip binary
    if source[..source.len().min(8192)].contains(&0) {
        return;
    }

    // Clear old chunks
    let _ = conn.execute(
        "DELETE FROM code_content_chunks WHERE project_id = ?1 AND file_path = ?2",
        rusqlite::params![project_id, &rel],
    );

    let lang = path.extension().map(|e| e.to_string_lossy().to_string());
    let chunks = chunker::chunk_file_content(&source, &rel, project_id, lang.as_deref());
    if !chunks.is_empty() {
        upsert_content_chunks(conn, &chunks);
    }

    let _ = meta; // used for size check above
}

/// Invalidate all index data for a project.
pub fn invalidate(conn: &Connection, project_id: &str) -> anyhow::Result<()> {
    conn.execute(
        "DELETE FROM code_symbols WHERE project_id = ?1",
        rusqlite::params![project_id],
    )?;
    conn.execute(
        "DELETE FROM code_indexed_files WHERE project_id = ?1",
        rusqlite::params![project_id],
    )?;
    conn.execute(
        "DELETE FROM code_content_chunks WHERE project_id = ?1",
        rusqlite::params![project_id],
    )?;
    eprintln!("Invalidated code index for project {project_id}");
    Ok(())
}

// ── SQLite helpers ─────────────────────────────────────────────────────

fn upsert_symbols(conn: &Connection, symbols: &[crate::models::Symbol]) {
    let now = iso_now();
    for sym in symbols {
        let _ = conn.execute(
            "INSERT INTO code_symbols (
                id, project_id, file_path, name, qualified_name,
                kind, language, byte_start, byte_end,
                line_start, line_end, signature, docstring,
                parent_symbol_id, content_hash, summary,
                created_at, updated_at
            ) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12,?13,?14,?15,?16,?17,?18)
            ON CONFLICT(id) DO UPDATE SET
                name=excluded.name, qualified_name=excluded.qualified_name,
                kind=excluded.kind, byte_start=excluded.byte_start,
                byte_end=excluded.byte_end, line_start=excluded.line_start,
                line_end=excluded.line_end, signature=excluded.signature,
                docstring=excluded.docstring, parent_symbol_id=excluded.parent_symbol_id,
                language=excluded.language, content_hash=excluded.content_hash,
                updated_at=excluded.updated_at",
            rusqlite::params![
                sym.id,
                sym.project_id,
                sym.file_path,
                sym.name,
                sym.qualified_name,
                sym.kind,
                sym.language,
                sym.byte_start as i64,
                sym.byte_end as i64,
                sym.line_start as i64,
                sym.line_end as i64,
                sym.signature,
                sym.docstring,
                sym.parent_symbol_id,
                sym.content_hash,
                sym.summary,
                &now,
                &now,
            ],
        );
    }
}

fn upsert_file(conn: &Connection, file: &IndexedFile) {
    let _ = conn.execute(
        "INSERT INTO code_indexed_files (
            id, project_id, file_path, language, content_hash,
            symbol_count, byte_size, indexed_at
        ) VALUES (?1,?2,?3,?4,?5,?6,?7,?8)
        ON CONFLICT(id) DO UPDATE SET
            content_hash=excluded.content_hash,
            symbol_count=excluded.symbol_count,
            byte_size=excluded.byte_size,
            indexed_at=excluded.indexed_at",
        rusqlite::params![
            file.id,
            file.project_id,
            file.file_path,
            file.language,
            file.content_hash,
            file.symbol_count as i64,
            file.byte_size as i64,
            file.indexed_at,
        ],
    );
}

fn upsert_content_chunks(conn: &Connection, chunks: &[crate::models::ContentChunk]) {
    for chunk in chunks {
        let _ = conn.execute(
            "INSERT INTO code_content_chunks (
                id, project_id, file_path, chunk_index,
                line_start, line_end, content, language, created_at
            ) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9)
            ON CONFLICT(id) DO UPDATE SET
                content=excluded.content,
                line_start=excluded.line_start,
                line_end=excluded.line_end",
            rusqlite::params![
                chunk.id,
                chunk.project_id,
                chunk.file_path,
                chunk.chunk_index as i64,
                chunk.line_start as i64,
                chunk.line_end as i64,
                chunk.content,
                chunk.language,
                chunk.created_at,
            ],
        );
    }
}

fn upsert_project_stats(conn: &Connection, project: &IndexedProject) {
    let _ = conn.execute(
        "INSERT INTO code_indexed_projects (
            id, root_path, total_files, total_symbols,
            last_indexed_at, index_duration_ms
        ) VALUES (?1,?2,?3,?4,?5,?6)
        ON CONFLICT(id) DO UPDATE SET
            total_files=excluded.total_files,
            total_symbols=excluded.total_symbols,
            last_indexed_at=excluded.last_indexed_at,
            index_duration_ms=excluded.index_duration_ms",
        rusqlite::params![
            project.id,
            project.root_path,
            project.total_files as i64,
            project.total_symbols as i64,
            project.last_indexed_at,
            project.index_duration_ms as i64,
        ],
    );
}

fn delete_file_data(conn: &Connection, project_id: &str, file_path: &str, neo4j: Option<&Neo4jClient>) {
    // Delete graph data first
    if let Some(client) = neo4j {
        crate::neo4j::delete_file_graph(client, project_id, file_path);
    }

    let _ = conn.execute(
        "DELETE FROM code_symbols WHERE project_id = ?1 AND file_path = ?2",
        rusqlite::params![project_id, file_path],
    );
    let _ = conn.execute(
        "DELETE FROM code_indexed_files WHERE project_id = ?1 AND file_path = ?2",
        rusqlite::params![project_id, file_path],
    );
    let _ = conn.execute(
        "DELETE FROM code_content_chunks WHERE project_id = ?1 AND file_path = ?2",
        rusqlite::params![project_id, file_path],
    );
}

fn get_stale_files(
    conn: &Connection,
    project_id: &str,
    current_hashes: &HashMap<String, String>,
) -> std::collections::HashSet<String> {
    let mut stale = std::collections::HashSet::new();

    // Create temp table for comparison
    let _ = conn.execute_batch(
        "CREATE TEMP TABLE IF NOT EXISTS _current_hashes \
         (file_path TEXT PRIMARY KEY, content_hash TEXT); \
         DELETE FROM _current_hashes;",
    );

    for (path, hash) in current_hashes {
        let _ = conn.execute(
            "INSERT INTO _current_hashes (file_path, content_hash) VALUES (?1, ?2)",
            rusqlite::params![path, hash],
        );
    }

    if let Ok(mut stmt) = conn.prepare(
        "SELECT ch.file_path FROM _current_hashes ch \
         LEFT JOIN code_indexed_files cf \
             ON cf.project_id = ?1 AND cf.file_path = ch.file_path \
         WHERE cf.file_path IS NULL OR cf.content_hash != ch.content_hash",
    ) {
        if let Ok(rows) = stmt.query_map(rusqlite::params![project_id], |row| {
            row.get::<_, String>(0)
        }) {
            for row in rows.flatten() {
                stale.insert(row);
            }
        }
    }

    let _ = conn.execute_batch("DROP TABLE IF EXISTS _current_hashes;");
    stale
}

fn get_orphan_files(
    conn: &Connection,
    project_id: &str,
    current_hashes: &HashMap<String, String>,
) -> Vec<String> {
    let mut orphans = Vec::new();
    if let Ok(mut stmt) = conn.prepare(
        "SELECT file_path FROM code_indexed_files WHERE project_id = ?1",
    ) {
        if let Ok(rows) = stmt.query_map(rusqlite::params![project_id], |row| {
            row.get::<_, String>(0)
        }) {
            for row in rows.flatten() {
                if !current_hashes.contains_key(&row) {
                    orphans.push(row);
                }
            }
        }
    }
    orphans
}

fn count_rows(conn: &Connection, table: &str, project_id: &str) -> usize {
    let sql = format!("SELECT COUNT(*) FROM {table} WHERE project_id = ?1");
    conn.query_row(&sql, rusqlite::params![project_id], |row| {
        row.get::<_, i64>(0)
    })
    .unwrap_or(0) as usize
}

fn relative_path(path: &Path, root: &Path) -> anyhow::Result<String> {
    let abs = path.canonicalize()?;
    let root_abs = root.canonicalize()?;
    Ok(abs
        .strip_prefix(&root_abs)?
        .to_string_lossy()
        .to_string())
}

fn iso_now() -> String {
    use std::time::SystemTime;
    let secs = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    format!("{secs}")
}
